lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/file/download/v2/DownloadListDAOImpl.java

```diff
@@ -163,10 +207,194 @@ public DBODownloadList getDBODownloadList(Long userId) {
+		// Create a temp table that contains only the IDs we want to keep.
+		Set<Long> idsToKeep = Arrays.stream(items).map(i -> KeyFactory.stringToKey(i.getFileEntityId()))
+				.collect(Collectors.toSet());
+		String tempTableName = createTemporaryTableOfAvailableFiles(
+				t -> t.stream().filter(i -> idsToKeep.contains(i)).collect(Collectors.toList()), userId, BATCH_SIZE);
+		try {
+			MapSqlParameterSource params = new MapSqlParameterSource();
+			params.addValue("principalId", userId);
+			params.addValue("depth", NodeConstants.MAX_PATH_DEPTH_PLUS_ONE);
+			String sql = String.format(DOWNLOAD_LIST_RESULT_TEMPLATE, tempTableName);
+			List<DownloadListItemResult> unorderedResults = namedJdbcTemplate.query(sql, params, RESULT_MAPPER);
+			
+			// Put the results in same order as the request. Note: O(n*m) where both n and m
+			// should be small.
+			return Arrays.stream(items).map(i -> unorderedResults.stream().filter(u -> isMatch(i, u)).findFirst().get())
+					.collect(Collectors.toList());
+		}finally {
+			dropTemporaryTable(tempTableName);
+		}
+	}
+
+	/**
+	 * Does the given DownloadListItem match the given DownloadListItemResult?
+	 * 
+	 * @param item
+	 * @param result
+	 * @return
+	 */
+	public static boolean isMatch(DownloadListItem item, DownloadListItemResult result) {
+		if (item.getFileEntityId().equals(result.getFileEntityId())) {
+			return Objects.equal(item.getVersionNumber(), result.getVersionNumber());
+		}
+		return false;
+	}
+
+	@Override
+	public List<DownloadListItemResult> getFilesAvailableToDownloadFromDownloadList(EntityAccessCallback accessCallback,
+			Long userId, List<Sort> sort, Long limit, Long offset) {
+		/*
+		 * The first step is to create a temporary table containing all of the entity
+		 * IDs from the user's download list that the user can download.
+		 */
+		String tempTableName = createTemporaryTableOfAvailableFiles(accessCallback, userId, BATCH_SIZE);
+		try {
+			StringBuilder sqlBuilder = new StringBuilder(String.format(DOWNLOAD_LIST_RESULT_TEMPLATE, tempTableName));
+			sqlBuilder.append(buildAvailableDownloadQuerySuffix(sort, limit, offset));
+			MapSqlParameterSource params = new MapSqlParameterSource();
+			params.addValue("principalId", userId);
+			params.addValue("depth", NodeConstants.MAX_PATH_DEPTH_PLUS_ONE);
+			params.addValue("limit", limit);
+			params.addValue("offset", offset);
+			return namedJdbcTemplate.query(sqlBuilder.toString(),params, RESULT_MAPPER);
+		}finally {
+			dropTemporaryTable(tempTableName);
+		}
+	}
+	
+	/**
+	 * Build the SQL suffix to handle both sorting and paging based on the provided sorting and paging.
+	 * @param sort
+	 * @param limit
+	 * @param offset
+	 * @return
+	 */
+	public static String buildAvailableDownloadQuerySuffix(List<Sort> sort, Long limit, Long offset) {
+		StringBuilder builder = new StringBuilder();
+		if(sort != null && !sort.isEmpty()) {
+			builder.append(" ORDER BY");
+			boolean first = true;
```

Comment:
Use a string joiner (or Collectors.joining() in Stream operations):

Something on the lines of: s.stream().map(mapping function to the columnname + desc/asc).collect(Collectors.joining(","));